feat(ast_tools): Add #[estree(append_to)], remove some custom serialization code

ottomated opened the pull request:
Removed custom logic for ObjectPattern, ArrayPattern, etc. in favor of a custom attribute macro `#[estree(append_to = "foo")]` as part of #6347

ottomated commented:
* **#7149** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7149?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ottomated and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_estree/src/lib.rs:
```diff
@@ -1,3 +1,25 @@
 pub trait ESTree {}
+
+#[cfg(feature = "serialize")]
+pub struct AppendTo<'a, TVec, TChild>(pub &'a [TVec], pub &'a Option<TChild>);
```
Personally, I generally prefer named fields over tuple structs. It's more verbose, but it makes it easier to read and understand the code.

`AppendTo` could also do with a comment explaining what it does. Oxc's codebase is not all as well-commented as it should be, but we're trying to improve in that department.

overlookmotel commented on npm/oxc-types/types.d.ts:
```diff
@@ -384,11 +384,13 @@ export type AssignmentTargetPattern = ArrayAssignmentTarget | ObjectAssignmentTa
 export interface ArrayAssignmentTarget extends Span {
   type: 'ArrayAssignmentTarget';
   elements: Array<AssignmentTargetMaybeDefault | AssignmentTargetRest | null>;
+  rest: AssignmentTargetRest | null;
```
Need to update TS type generation to take `append_to` into into account.

overlookmotel reviewed (approved):
Great!

Am merging this now, but @ottomated would you be able to address the 2 bits of feedback below in follow-on PRs? In particular, we should try to fix the TS types as soon as possible before we next make a release.

overlookmotel commented:
### Merge activity

* **Nov 6, 3:43 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

ottomated committed:
feat(ast_tools): Add #[estree(append_to)], remove some custom serialization code (#7149)

Removed custom logic for ObjectPattern, ArrayPattern, etc. in favor of a custom attribute macro `#[estree(append_to = "foo")]` as part of #6347

overlookmotel commented:
@ottomated Just to let you know I'm away until next Tues, so have merged these 2 PRs and left notes for suggested follow-on changes, rather than requesting changes now and blocking your progress.

If you want to continue onwards in my absence, please just stack up the PRs, and I'll review them when I get back.

If you're able to fix the broken TS types for the `rest` fields, please ask Boshen to merge that one. If it's OK with you, I'd prefer to review any further PRs myself, as I think I'm more aware of the context.
